lua-4.1-work3/doc/luac.1
.\" luac.man,v 1.22 2001/07/19 20:06:36 lhf Exp
.TH LUAC 1 "2001/07/19 20:06:36"
.SH NAME
luac \- Lua compiler
.SH SYNOPSIS
.B luac
[
.I options
] [
.I filenames
]
.SH DESCRIPTION
.B luac
is the Lua compiler.
It translates programs written in the Lua programming language
into binary files that can be loaded and executed with
.B lua_dofile
in C or with
.B dofile
in Lua
(or loaded with
.B lua_loadfile
in C or with
.B loadfile
in Lua).
.LP
The main advantages of precompiling chunks are:
faster loading,
protecting source code from user changes,
and
off-line syntax error detection.
.LP
Pre-compiling does not imply faster execution
because in Lua chunks are always compiled into bytecodes before being executed.
.B luac
simply allows those bytecodes to be saved in a file for later execution.
.LP
.B luac
produces a single output file containing the bytecodes
for all source files given.
By default,
the output file is named
.BR luac.out ,
but you can change this with the
.B \-o
option.
.LP
The binary files created by
.B luac
are portable to all architectures with the same word size.
This means that
binary files created on a 32-bit platform (such as Intel)
can be read without change in another 32-bit platform (such as Sparc),
even if the byte order (``endianness'') is different.
On the other hand,
binary files created on a 16-bit platform cannot be read in a 32-bit platform,
nor vice-versa.
.LP
In the command line,
you can mix
text files containing Lua source and
binary files containing precompiled chunks.
This is useful to combine several precompiled chunks,
even from different (but compatible) platforms,
into a single precompiled chunk.
.LP
You can use
.B "\-"
to indicate the standard input as a source file.
.LP
The internal format of the binary files produced by
.B luac
is likely to change when a new version of Lua is released.
So,
save the source files of all Lua programs that you precompile.
.LP
.SH OPTIONS
Options must be separate.
.TP
.B \-l
produce a listing of the compiled bytecode for Lua's virtual machine.
Listing bytecodes is useful to learn about Lua's virtual machine.
If no files are given, then
.B luac
loads
.B luac.out
and lists its contents.
.TP
.BI \-o " file"
output to
.IR file ,
instead of the default
.BR luac.out .
The output file may be a source file because
all files are loaded before the output file is written.
Be careful not to overwrite precious files.
.TP
.B \-p
load files but do not generate any output file.
Used mainly for syntax checking and for testing precompiled chunks:
corrupted files will probably generate errors when loaded.
Lua always performs a thorough integrity test on precompiled chunks.
Bytecode that passes this test is completely safe,
in the sense that it will not break the interpreter.
However,
there is no guarantee that such code does anything sensible.
(None can be given, because the halting problem is unsolvable.)
If no files are given, then
.B luac
loads
.B luac.out
and tests its contents.
No messages are displayed if the file passes the integrity test.
.TP
.B \-s
strip debug information before writing the output file.
This saves some space in very large chunks,
but if errors occur when running these chunks,
then the error messages may not contain the full information they usually do
(line numbers and names of locals are lost).
.TP
.B \-v
print version information.
.SH FILES
.TP 15
.B luac.out
default output file
.SH "SEE ALSO"
.BR lua (1)
.br
http://www.lua.org/
.SH DIAGNOSTICS
Error messages should be self explanatory.
.SH AUTHORS
L. H. de Figueiredo,
R. Ierusalimschy and
W. Celes
(lua@tecgraf.puc-rio.br)
.\" EOF


ChangeLog
2002-01-23  Peter Wang  <tjaden@users.sourceforge.net>

	* mylua.c: Seed lua random() function.

	* gamesrv.c: Changed the test for a very small yv from
	0 <= yv < 0.0000001 to 0 <= yv < 0.0001.

	* basic-player.lua: Added rest of death animations.

2002-01-22  Peter Wang  <tjaden@users.sourceforge.net>

	* basic-powerup.lua, basic-player.lua:
	Made lightamp powerup give player a big blue light.

	* basic-player.lua: Made local player's light turn red if gotten
	shot, and made corpses use the red light.

	* basic-player.lua:
	Added start of backpacks.  Also minor clean up for walking anim.

	* editor-init.lua:
	Some functions cannot be run while in the editor or it will crash, so
	we make them into dummy functions.
	
	* init.lua, editor/editor.c:
	Only run editor-init.lua when actually running the editor.
	This allows us to put in editor-specific code in there.

	* gamesrv.c, gameclt.c, basic-weapon.lua, init.lua:
	Added object type aliases.

	* gamesrv.c: Added network traffic monitoring.
	* textface.c, gamesrv.h: Added set_status() method to
	game_server_interface.

	* basic-player.lua, basic-powerup.lua:
	Made health items give health.

	* basic-weapon.lua, basic-powerup.lua, basic-base.lua:
	Allow a second argument to Objtype and Weapon, which should
	contain a table of fields to add into or override in the first
	argument.

	* basic-weapon.lua, basic-player.lua: Animated weapon firing.

	* netmsg.h, gamesrv.c, gameclt.c, basic-weapon.lua,
	* basic-player.lua: Added weapon switching with keyboard.

	* basic-weapon.lua, basic-player.lua: Added ammo limits.

2002-01-21  Peter Wang  <tjaden@users.sourceforge.net>

	* basic-player.lua: Made player bleed when on low health.

	* gameclt.h, bindgen.lua, basic-weapon.lua, basic-player.lua,
	* camera.h, camera.c, gameclt.c:
	Made switching to the sniper rifle increase the max distance the
	camera can see.

	* basic-weapon.lua: Changed the generalised declarators to take
	table arguments, which are more verbose and easier to read.

	* basic-player.lua, basic-weapon.lua:
	Use the right image for the weapon we are holding (not animated yet).

	* gameclt.c, bindgen.lua, gamesrv.h, gamesrv.c, netmsg.h:
	Added mechanism for server scripts to tell clients to execute object
	methods (i.e. remote calls).

	* object.h, object.c:
	Allowed object_call() to take an nargs argument.  The arguments should
	be already be pushed on the lua stack.

	* basic-base.lua: Added merge function.

	* gamesrv.c: Fixed a bug where sometimes you couldn't jump because
	the player's yv was very, very, very close to zero but not zero.

2002-01-20  Peter Wang  <tjaden@users.sourceforge.net>

	* gameclt.c, camera.h, camera.c:
	Added ability to "push" the camera, i.e. for the sniper rifle so you
	can see further that your mouse would normally allow on the screen.
	This isn't actually enabled when you pick up a sniper rifle, yet.

	* basic-weapon.lua:
	Added initial code for AK-47 and miniguns, and shared a lot of code
	between the weapon types.

	* basic-tilelike.lua:
	Some initial code to make crates destructible.

	* basic-powerup.lua, basic-base.lua:
	Added hide_and_respawn_later() method which is used a lot by objects.

2002-01-19  Peter Wang  <tjaden@users.sourceforge.net>

	* objgen.lua: Exposed object's id value.  This was done for corpse
	tracking after death, but I'm not too sure it should be done like
	this generally.

	* gamesrv.h, gamesrv.c, bindgen.lua: Made spawn_object,
	spawn_projectile return the objects they create.

	* gameclt.c, basic-player.lua:
	Added tracking corpses after death.

	* basic-weapon.lua:
	The start of some new weapons: shotguns and sniper rifles.

	* bindgen.lua, gamesrv.h, gamesrv.c:
	Added a delta_angle parameter to game_server_spawn_projectile().

	* object.c: Fixed collision detection so that _both_ objects'
	conditions for collision detection have to be passed before
	checking is done.

	* editor/mdobject.c, bindgen.lua:
	Allow objects without a category which don't show up in the editor.
	* blod.c: Made blod types not show up in map editor.
	* basic-player.lua: Made test corpse not show up in editor.

	* objtypes.h, objtypes.c, basic-base.lua:
	Return error code if registering an object type failed.

	* object.c:
	Slow down player if ramped movement used (walking up stairs, etc.)

	* blod.c: Added remaining blods and gave them variable masses and
	timeouts.

	* gamesrv.c: Link players at top of object list so that they
	appear above objects that were present at map load time (ladders,
	barrels, etc.)

	* gamesrv.c: Made jumping from standing on top of ladder position
	to be as high as jump from normal tile.

	* basic-player.lua: Decreased player xv_decay, yv_decay to
	slow down movement a little.

	* store/README, store/CHANGES, store/hash.c, store/store.c:
	Reverted to the libc string functions and only support
	ASCII/UTF-8 (faster).

	* basic-player.lua, bindgen.lua, gameclt.c, gameinit.c,
	* gamesrv.c, gamesrv.h, netmsg.h, blod.c, blod.h:
	Added blods.

	* object.c, object.h: Added client-processed objects.

2002-01-18  Peter Wang  <tjaden@users.sourceforge.net>

	* blood.c: Fixed a bug where it would try to allocate zero new
	particles.

	* editor/mdobject.c: Added snap to grid.

	* gamesrv.c:
	Got rid of a bad optimisation which broke down the mechanism which
	stopped players falling into ladders.

	* editor/mdobject.c:
	Added different cursors for different object editting actions.

	* editor/cursor.c:
	Created dot cursor bitmap at module initialisation instead of
	creating it everytime it was set to improve efficiency.

	* gui/guiwmdef.c, gui/gui.c:
	Added display switching in callback to force redrawing the screen.

	* editor/cursor.c:
	Created dot cursor bitmap at module initialisation instead of
	creating it every time it was set to improve efficiency.

	* editor/mdobject.c:
	Added different cursors for different object editting actions.

	* object.c:
	Use rawget, rawset instead of gettable, settable in places.


lua-4.1-work3/src/luac/luac.c
/*
** $Id: luac.c,v 1.33 2001/11/29 01:00:34 lhf Exp $
** Lua compiler (saves bytecodes to files; also list bytecodes)
** See Copyright Notice in lua.h
*/

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "luac.h"
#include "lfunc.h"
#include "lmem.h"
#include "lobject.h"
#include "lopcodes.h"
#include "lstring.h"
#include "lundump.h"

#define	OUTPUT	"luac.out"		/* default output file */

static void usage(const char* message, const char* arg);
static int doargs(int argc, const char* argv[]);
static Proto* load(const char* filename);
static Proto* combine(Proto** P, int n);
static void strip(Proto* f);
static void cannot(const char* name, const char* what, const char* mode);
static void fatal(const char* message);

static lua_State* L=NULL;
static int listing=0;			/* list bytecodes? */
static int dumping=1;			/* dump bytecodes? */
static int stripping=0;			/* strip debug information? */
static const char* output=OUTPUT;	/* output file name */

/* need this to handle errors during parse or undump */
static int ERRORMESSAGE(lua_State *l)
{
 fatal(lua_tostring(l,1));
 return 0;
}

#define	IS(s)	(strcmp(argv[i],s)==0)

int main(int argc, const char* argv[])
{
 Proto** P,*f;
 int i=doargs(argc,argv);
 argc-=i; argv+=i;
 if (argc<=0) usage("no input files given",NULL);
 L=lua_open(0);
 lua_register(L,LUA_ERRORMESSAGE,ERRORMESSAGE);
 P=luaM_newvector(L,argc,Proto*);
 for (i=0; i<argc; i++)
  P[i]=load(IS("-")? NULL : argv[i]);
 f=combine(P,argc);
 if (listing) luaU_printchunk(f);
 if (dumping)
 {
  FILE* D=fopen(output,"wb");
  if (D==NULL) cannot(output,"open","out");
  if (stripping) strip(f);
  luaU_dumpchunk(f,D);
  if (ferror(D)) cannot(output,"write","out");
  fclose(D);
 }
 return 0;
}

static void usage(const char* message, const char* arg)
{
 if (message!=NULL)
 {
  fprintf(stderr,"luac: "); fprintf(stderr,message,arg); fprintf(stderr,"\n");
 }
 fprintf(stderr,
 "usage: luac [options] [filenames].  Available options are:\n"
 "  -        process stdin\n"
 "  -l       list\n"
 "  -o file  output file (default is \"" OUTPUT "\")\n"
 "  -p       parse only\n"
 "  -s       strip debug information\n"
 "  -v       show version information\n"
 );
 exit(1);
}

static int doargs(int argc, const char* argv[])
{
 int i;
 for (i=1; i<argc; i++)
 {
  if (*argv[i]!='-')			/* end of options */
   break;
  else if (IS("-"))			/* end of options; use stdin */
   return i;
  else if (IS("-l"))			/* list */
   listing=1;
  else if (IS("-o"))			/* output file */
  {
   output=argv[++i];
   if (output==NULL) usage(NULL,NULL);
  }
  else if (IS("-p"))			/* parse only */
   dumping=0;
  else if (IS("-s"))			/* strip debug information */
   stripping=1;
  else if (IS("-v"))			/* show version */
  {
   printf("%s  %s\n",LUA_VERSION,LUA_COPYRIGHT);
   if (argc==2) exit(0);
  }
  else					/* unknown option */
   usage("unrecognized option `%s'",argv[i]);
 }
 if (i==argc && (listing || !dumping))
 {
  dumping=0;
  argv[--i]=OUTPUT;
 }
 return i;
}

static Proto* load(const char* filename)
{
 switch (lua_loadfile(L,filename))
 {
  case 0:
  if (errno!=0) cannot(filename,"read","in");
  {
   const Closure* c=lua_topointer(L,-1);
   return c->l.p;
  }
  case LUA_ERRFILE:
   cannot(filename,"open","in");
   break;
  case LUA_ERRSYNTAX:
   fatal("syntax error");
   break;
  case LUA_ERRRUN:
   fatal("run-time error");
   break;
  case LUA_ERRMEM:
   fatal("not enough memory");
   break;
  case LUA_ERRERR:
   fatal("error in error handling");
   break;
  default:
   fatal("unknown status returned by lua_loadfile");
   break;
 }
 return NULL;
}

static Proto* combine(Proto** P, int n)
{
 if (n==1)
  return P[0];
 else
 {
  int i,pc=0;
  Proto* f=luaF_newproto(L);
  f->source=luaS_new(L,"=(luac)");
  f->maxstacksize=1;
  f->p=P;
  f->sizep=n;
  f->sizecode=2*n+1;
  f->code=luaM_newvector(L,f->sizecode,Instruction);
  for (i=0; i<n; i++)
  {
   f->code[pc++]=CREATE_ABc(OP_CLOSURE,0,i);
   f->code[pc++]=CREATE_ABC(OP_CALL,0,0,0);
  }
  f->code[pc++]=CREATE_ABC(OP_RETURN,0,0,0);
  return f;
 }
}

static void strip(Proto* f)
{
 int i,n=f->sizep;
 f->lineinfo=NULL;
 f->sizelineinfo=0;
 f->source=luaS_new(L,"=(none)");
 f->locvars=NULL;
 f->sizelocvars=0;
 for (i=0; i<n; i++) strip(f->p[i]);
}

static void cannot(const char* name, const char* what, const char* mode)
{
 fprintf(stderr,"luac: cannot %s %sput file ",what,mode);
 perror(name);
 exit(1);
}

static void fatal(const char* message)
{
 fprintf(stderr,"luac: %s\n",message);
 exit(1);
}

/*
* the code below avoids the parsing modules (lcode, llex, lparser).
* it is useful if you only want to load binary files.
* this works for interpreters like lua.c too.
*/

#ifdef NOPARSER

#include "llex.h"
#include "lparser.h"
#include "lzio.h"

void luaX_init(lua_State *L) {
  UNUSED(L);
}

Proto *luaY_parser(lua_State *L, ZIO *z) {
  UNUSED(z);
  lua_error(L,"parser not loaded");
  return NULL;
}

#endif
